var/spack/repos/builtin/packages/ffb/package.py

```diff
@@ -0,0 +1,133 @@
+# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+import os
+from spack import *
+
+
+class Ffb(MakefilePackage):
+    """Computational Fluid Dynamics Software for aeroacoustics"""
+
+    homepage = "http://www.ciss.iis.u-tokyo.ac.jp/dl/index.php"
+    url      = "file://{0}/FrontFlow_blue.8.1.tar.gz".format(os.getcwd())
+    version('8.1', sha256='1ad008c909152b6c27668bafbad820da3e6ec3309c7e858ddb785f0a3d6e43ae')
+
+    patch('revocap_refiner.patch')
+
+    depends_on('mpi')
+    depends_on('blas')
+    depends_on('scalapack')
+
+    parallel = False
+
+    def edit(self, spec, prefix):
+        workdir = os.getcwd()
+        cflags = '-O3'
+        cxxflags = ['-O2', self.compiler.cxx_pic_flag]
+        fflags = '-O3 -mcmodel=large'
```

Comment:
Here's a suggestion for using a `flag_handler` method to only add default optimization levels if the user did not already set an optimization level:

```suggestion
    def flag_handler(self, name, flags):
        opt_flag_found = any(f in self.compiler.opt_flags for f in flags)
        if name == 'cflags':
            if not opt_flag_found:
                flags.append('-O3')
        elif name == 'cxxflags':
            if not opt_flag_found:
                flags.append('-O2')
            flags.append(self.compiler.cxx_pic_flag)
        if name == 'fflags':
            if not opt_flag_found:
                flags.append('-O3')
            flags.append('-mcmodel=large')
        if name in ('cflags', 'cxxflags', 'fflags'):
            return (None, flags, None)
        else:
            return (flags, None, flags)

    def edit(self, spec, prefix):
        workdir = os.getcwd()
        cflags = env['CFLAGS']
        cxxflags = env['CXXFLAGS']
        fflags = env['FFLAGS']
```

What this does:
1. If the user didn't put an optimization flag (-O*) in cflags, add `-O3`
2. If the user didn't put an optimization flag (-O*) in cxxflags, add `-O2`. No matter what, add the cxx pic flag to cxxflags. 
3. If the user didn't put an optimization flag (-O*) in fflags, add `-O3`. No matter what, add `-mcmodel=large` to fflags.
4. Tell Spack to put cflags, cxxflags, and fflags in the canonical environment variables so you can access them later.
5. Tell Spack to put other flags (ldflags, ldlibs) in the Spack variables so they will be injected (basically, don't do anything with them in the build system and let Spack handle it.

Then in the edit method, I can get the flags from the environment variables. This would require changing the parts that deal with cxxflags lower down to use a string instead of a list.

Reply:
Thank you for the detailed explanation.
I now understand how to use flag_handler.